contributors/design-proposals/dynamic-admission-control-configuration.md

```diff
@@ -0,0 +1,354 @@
+admission controller` and the `generic webhook admission controller` are the two
+controllers that set default initializers and external admission hooks for
+resources newly created. These two admission controllers are in the same binary
+as the apiserver. This
+[section](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/admission_control_extension.md#dynamic-configuration)
+gave a preliminary design of the dynamic configuration of the list of the
+default admission controls. This document hashes out the implementation details.
+
+## Goals
+
+* Admin is able to predict what initializers/webhooks will be applied to newly
+  created objects.
+
+* Admin needs to be able to ensure initializers/webhooks config will be applied within some bound
+
+* As a fallback, admin can always restart an apiserver and guarantee it sees the latest config
+
+* Do not block the entire cluster if the intializers/webhooks are not ready
+  after registration.
+
+## Specification
+
+We assume initializers could be "fail open". We need to update the extensible
+admission control
+[proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/admission_control_extension.md)
+if this is accepted.
+
+The schema is evolved from the prototype in
+[#132](https://github.com/kubernetes/community/pull/132).
+
+```golang
+// AdmissionControlConfiguration describes the configuration of intializers and
+// external admission webhooks.
+type AdmissionControlConfiguration struct {
+    metav1.TypeMeta
+
+    v1.ObjectMeta
+
+    // Initializers is a list of resources and their default initializers
+    // Order-sensitive.
+    // When merging multiple AdmissionControlConfigurations, we sort the intializers
+    // from different AdmissionControlConfigurations by the name of the
+    // AmdissionControlConfigurations; the order of the intializers from the same
+    // AdmissionControlConfiguration is preserved.
+    // +optional
+    Initializers []Initializer `json:"initializers,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+
+    // ExternalAdmissionHooks is a list of external admission webhooks and the
+    // affected resources and operations.
+    // +optional
+    ExternalAdmissionHooks []ExternalAdmissionHook `json:"externalAdmissionHooks,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// Initializer describes the name and the failure policy of an initializer, and
+// what resources it applies to.
+type Initializer struct {
+    // Name is the identifier of the initializer. It will be added to the
+    // object that needs to be initialized.
+    // Required
+    Name string `json:"name"`
+
+    // Rules describes what resources/subresources the initializer cares about.
+    // The intializer cares about an operation if it matches _any_ Rule.
+    // Rule.Verbs must be string{"CREATE"}, otherwise it is a validation error.
```

Comment:
I'm not sure about this bit. It's ugly to expose an API of `[]string` that is only valid with one value. One alternative: 
```golang
Rule {
   APIGroups []string
   APIVersions []string
   ...
}

WebhookRule {
    Verbs []string
    Rule `json:",inline"`
}
```
Is embedding and inlining good?